### ISSUE TITLE
"export * from something" not working in specific case

### ISSUE THREAD

**Issue by jdanyow** [systemjs/systemjs#858]
Aurelia's `framework` module [exports everything](https://github.com/aurelia/framework/blob/0.17.0/src/index.js#L3) in the `aurelia-metadata` module.  The compiled code looks like [this](https://github.com/jdanyow/skeleton-navigation/blob/gh-pages/jspm_packages/github/aurelia/framework%400.17.0/aurelia-framework.js#L386).

We're seeing a strange issue where the export doesn't work for this particular module.  To see this issue in action, open chrome's debug tools and go [here](http://jdanyow.github.io/skeleton-navigation/).

You'll break [here](https://github.com/jdanyow/skeleton-navigation/blob/gh-pages/index.html#L21) in some code I added that detects the issue as it occurs (`DecoratorApplicator` is one of `aurelia-metadata`'s exports).  This piece of code isn't required to reproduce the issue, it was just my attempt to figure out when and why it was happening.

The real issue is [here](https://github.com/jdanyow/skeleton-navigation/blob/gh-pages/src/main.js#L6).

cc @EisenbergEffect

**Comment by guybedford** [comment]
@jdanyow thanks for sharing the detailed information, it helps a lot to have some background. I just ran the skeleton and saw the logs as you described.

The redefinition issue may be since `DecoratorApplicator` is imported by the same name from two different modules? Note that that is strictly not supported in ES modules. I'd suggest rescoping one of the names locally via `import {DecoratorApplicator as MetadataDecoratorApplicator} from 'aurelia-metadata'` for example.

As for the undefined issue, these are usually due to circular references.

Please do let me know if I'm missing something or what to check for though further.

**Comment by jdanyow** [comment]
Thanks for taking a look at this- here's some more info:

The `main.js` file that demonstrates the issue has the following code:

```javascript
import 'bootstrap';
import {DecoratorApplicator, ObserverLocator} from 'aurelia-framework';
import {DecoratorApplicator as Test} from 'aurelia-metadata';

if (!DecoratorApplicator && ObserverLocator && Test) {
  console.error(`DecoratorApplicator isn't exported from 'aurelia-framework'?`);
  debugger;
}
```

Really all that is required to reproduce the issue is the following:
```javascript
import {DecoratorApplicator} from 'aurelia-framework';

if (!DecoratorApplicator) {
  console.error(`DecoratorApplicator isn't exported from 'aurelia-framework'?`);
  debugger;
}
```

All the other stuff was either
- irrelevant (importing bootstrap),
- intended to prove that `aurelia-framework` was able to re-export stuff (the import of ObserverLocator and subsequent check for it's existence proves that `aurelia-framework` was able to re-export ObserverLocator from `aurelia-binding`),
- or intended to prove that `DecoratorApplicator` was in fact exported from `aurelia-metadata` and in turn should have been exported from `aurelia-framework`.

In other words, why is aurelia-framework able to re-export the aurelia-binding's components but not aurelia-metadata's?

**Comment by guybedford** [comment]
@jdanyow I had a deeper look at the debugging of this and `aurelia-metadata` seemed to export fine, but in `aurelia-framework` the export * seemed to not work properly on this line - https://github.com/jdanyow/skeleton-navigation/blob/gh-pages/jspm_packages/github/aurelia/framework%400.17.0/aurelia-framework.js#L386.

Basically this function is supposed to copy all values from aurelia metadata onto a new empty object. If you follow that in the debugger, and expand the code to step through the `_defaults` function, you can see it iterate over `DecoratorApplicator` in the main loop at https://github.com/jdanyow/skeleton-navigation/blob/gh-pages/jspm_packages/github/aurelia/framework%400.17.0/aurelia-framework.js#L9 when starting the debugger from https://github.com/jdanyow/skeleton-navigation/blob/gh-pages/jspm_packages/github/aurelia/framework%400.17.0/aurelia-framework.js#L386.

*Then it ignores DecoratorApplicator* on this code in that line - `if (value && value.configurable && obj[key] === undefined)`. The reason for this is that the `DecoratorApplicator` export is not configurable.

Altering that statement to `if (value && obj[key] === undefined)` seems to fix the issue as far as I can tell.

The above sounds to me like a Babel bug, although there may be things we can do in SystemJS too to work around it. I would need to investigate further to be able to be sure to report it further though.

**Comment by jdanyow** [comment]
I see, so checking the `configurable` attribute of the property descriptor is potentially unnecessary and could be a babel issue.  Should I open an issue in babel for this?

Why does aurelia's aurelia-metadata module land in this SystemJS code while other aurelia modules don't (and ultimately don't run into the babel issue):
![stack](http://i.imgur.com/aSG3dpk.png)

Is there something we could be doing with the aurelia-metadata module to avoid this block of code?

**Comment by guybedford** [comment]
@jdanyow I've posted the issue to Babel. The workaround here would be to ensure `aurelia-metadata` is itself compiled into CommonJS and not run from ES module source.

**Comment by jdanyow** [comment]
Thanks @guybedford! - will keep an eye on the babel issue.  Closing this.
